File: python/sharepoint_folder_size.py
#!/usr/bin/env python3
"""
SharePoint Folder Size Calculator - Final Working Version
Uses Microsoft Graph API with proper folder path handling
"""

import os
import sys
import json
import csv
import argparse
import time
from datetime import datetime
from typing import Dict, List, Optional, Any, Tuple
import requests
from urllib.parse import quote


class GraphClient:
    """Microsoft Graph API client for SharePoint access"""
    
    def __init__(self, tenant_id: str, client_id: str, client_secret: str):
        self.tenant_id = tenant_id
        self.client_id = client_id
        self.client_secret = client_secret
        self.access_token = None
        self.headers = {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }
        
    def authenticate(self) -> bool:
        """Authenticate and get access token"""
        print("\n" + "="*50)
        print("AUTHENTICATING WITH MICROSOFT GRAPH")
        print("="*50)
        
        # Token endpoint
        token_url = f"https://login.microsoftonline.com/{self.tenant_id}/oauth2/v2.0/token"
        
        # Request body for Graph API
        data = {
            'grant_type': 'client_credentials',
            'client_id': self.client_id,
            'client_secret': self.client_secret,
            'scope': 'https://graph.microsoft.com/.default'
        }
        
        try:
            print("Requesting access token...")
            response = requests.post(token_url, data=data)
            response.raise_for_status()
            
            token_data = response.json()
            self.access_token = token_data.get('access_token')
            
            if self.access_token:
                self.headers['Authorization'] = f'Bearer {self.access_token}'
                print("✓ Authentication successful!")
                return True
            else:
                print("✗ No access token received")
                return False
                
        except requests.exceptions.RequestException as e:
            print(f"✗ Authentication error: {str(e)}")
            if hasattr(e, 'response') and e.response is not None:
                print(f"Response: {e.response.text}")
            return False
            
    def get_site_and_drive(self, site_url: str) -> Tuple[Optional[str], Optional[str], Optional[str]]:
        """Get the site ID and document library drive ID"""
        # Extract site path from URL
        parts = site_url.replace('https://', '').split('/')
        hostname = parts[0]
        site_path = '/'.join(parts[1:])
        
        # Get site
        api_url = f"https://graph.microsoft.com/v1.0/sites/{hostname}:/{site_path}"
        
        try:
            response = requests.get(api_url, headers=self.headers)
            if response.status_code == 200:
                site_data = response.json()
                site_id = site_data.get('id')
                site_name = site_data.get('displayName', 'Unknown')
                print(f"✓ Connected to site: {site_name}")
                
                # Get drives
                drives_url = f"https://graph.microsoft.com/v1.0/sites/{site_id}/drives"
                drives_response = requests.get(drives_url, headers=self.headers)
                
                if drives_response.status_code == 200:
                    drives_data = drives_response.json()
                    drives = drives_data.get('value', [])
                    
                    # Find the document library
                    for drive in drives:
                        if drive.get('name') in ['Documents', 'Shared Documents']:
                            return site_id, drive.get('id'), drive.get('name')
                    
                    # Return first drive if specific one not found
                    if drives:
                        return site_id, drives[0].get('id'), drives[0].get('name')
                        
            return None, None, None
        except Exception as e:
            print(f"✗ Error getting site: {str(e)}")
            return None, None, None
            
    def get_drive_item_by_path(self, site_id: str, drive_id: str, item_path: str) -> Optional[Dict]:
        """Get a drive item (file or folder) by its path"""
        if not item_path or item_path == '/':
            # Root folder
            api_url = f"https://graph.microsoft.com/v1.0/sites/{site_id}/drives/{drive_id}/root"
        else:
            # Specific path
            encoded_path = quote(item_path)
            api_url = f"https://graph.microsoft.com/v1.0/sites/{site_id}/drives/{drive_id}/root:/{encoded_path}"
            
        try:
            response = requests.get(api_url, headers=self.headers)
            if response.status_code == 200:
                return response.json()
            return None
        except Exception:
            return None
            
    def get_folder_children(self, site_id: str, drive_id: str, folder_id: str) -> Tuple[List[Dict], List[Dict]]:
        """Get children of a folder using its ID"""
        api_url = f"https://graph.microsoft.com/v1.0/sites/{site_id}/drives/{drive_id}/items/{folder_id}/children"
        
        files = []
        folders = []
        
        try:
            # Handle pagination
            while api_url:
                response = requests.get(api_url, headers=self.headers)
                if response.status_code == 200:
                    data = response.json()
                    items = data.get('value', [])
                    
                    for item in items:
                        if 'folder' in item:
                            # Skip system folders
                            if not (item['name'].startswith('_') or item['name'] == 'Forms'):
                                folders.append(item)
                        elif 'file' in item:
                            files.append(item)
                    
                    # Check for next page
                    api_url = data.get('@odata.nextLink')
                else:
                    break
                    
        except Exception as e:
            print(f"  ✗ Error getting folder items: {str(e)}")
            
        return files, folders
        
    def calculate_folder_size(self, site_id: str, drive_id: str, folder_item: Dict, depth: int = 0) -> Dict[str, Any]:
        """Calculate folder size recursively using folder item"""
        indent = "  " * depth
        folder_name = folder_item.get('name', 'Unknown')
        folder_id = folder_item.get('id')
        
        print(f"{indent}📁 {folder_name}")
        
        result = {
            'id': folder_id,
            'path': folder_item.get('webUrl', ''),
            'name': folder_name,
            'total_size': 0,
            'file_count': 0,
            'folder_count': 0,
            'files': [],
            'subfolders': []
        }
        
        if not folder_id:
            return result
            
        try:
            files, folders = self.get_folder_children(site_id, drive_id, folder_id)
            
            # Process files
            for file in files:
                file_info = {
                    'name': file.get('name', ''),
                    'size': file.get('size', 0),
                    'last_modified': file.get('lastModifiedDateTime', ''),
                    'path': file.get('webUrl', '')
                }
                result['files'].append(file_info)
                result['total_size'] += file_info['size']
                result['file_count'] += 1
                
                size_str = self.format_size(file_info['size'])
                print(f"{indent}  📄 {file_info['name']} ({size_str})")
            
            # Process subfolders
            result['folder_count'] = len(folders)
            for folder in folders:
                # Recursively process subfolder
                subfolder_result = self.calculate_folder_size(
                    site_id, drive_id, folder, depth + 1
                )
                result['subfolders'].append(subfolder_result)
                
                # Add subfolder totals to parent
                result['total_size'] += subfolder_result['total_size']
                result['file_count'] += subfolder_result['file_count']
                result['folder_count'] += subfolder_result['folder_count']
                
        except Exception as e:
            print(f"{indent}✗ Error processing folder: {str(e)}")
            
        return result
        
    def format_size(self, size_in_bytes: int) -> str:
        """Convert bytes to human-readable format"""
        for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
            if size_in_bytes < 1024.0:
                return f"{size_in_bytes:.2f} {unit}"
            size_in_bytes /= 1024.0
        return f"{size_in_bytes:.2f} PB"


class FolderSizeCalculator:
    """Main application class"""
    
    def __init__(self, client: GraphClient):
        self.client = client
        
    def analyze_site(self, site_url: str, folder_path: str) -> Optional[Dict[str, Any]]:
        """Analyze a folder in a SharePoint site"""
        # Get site and drive IDs
        site_id, drive_id, drive_name = self.client.get_site_and_drive(site_url)
        if not site_id or not drive_id:
            print("✗ Failed to get site or document library")
            return None
            
        print(f"✓ Using document library: {drive_name}")
        
        # Clean up folder path
        if folder_path:
            # Remove site prefix if present
            if folder_path.startswith('/sites/'):
                parts = folder_path.split('/')
                folder_path = '/'.join(parts[3:]) if len(parts) > 3 else ''
            
            # Remove leading/trailing slashes
            folder_path = folder_path.strip('/')
            
        print(f"\n🔍 Starting folder analysis...")
        print("="*50)
        start_time = time.time()
        
        # Get the folder item
        if not folder_path or folder_path in ['', 'Shared Documents', 'Documents']:
            # Analyze root of document library
            folder_item = self.client.get_drive_item_by_path(site_id, drive_id, '')
        else:
            # Specific folder
            folder_item = self.client.get_drive_item_by_path(site_id, drive_id, folder_path)
            
        if not folder_item:
            print(f"✗ Folder not found: {folder_path}")
            return None
            
        # Calculate folder size
        result = self.client.calculate_folder_size(site_id, drive_id, folder_item)
        
        end_time = time.time()
        print(f"\n⏱️  Analysis completed in {end_time - start_time:.2f} seconds")
        
        return result
        
    def print_summary(self, result: Dict[str, Any]):
        """Print analysis summary"""
        if not result:
            return
            
        print("\n" + "="*60)
        print("FOLDER SIZE SUMMARY")
        print("="*60)
        print(f"Folder: {result['name']}")
        print(f"Total Size: {self.client.format_size(result['total_size'])} ({result['total_size']:,} bytes)")
        print(f"Files: {result['file_count']:,}")
        print(f"Folders: {result['folder_count']:,}")
        print("="*60)
        
        # Get top 10 largest files
        all_files = self._get_all_files(result)
        all_files.sort(key=lambda x: x['size'], reverse=True)
        
        if all_files:
            print("\nTop 10 Largest Files:")
            print("-" * 60)
            for i, file in enumerate(all_files[:10], 1):
                print(f"{i:2d}. {file['name']:40s} {self.client.format_size(file['size']):>10s}")
                
    def _get_all_files(self, result: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Get all files recursively"""
        files = result['files'].copy()
        for subfolder in result['subfolders']:
            files.extend(self._get_all_files(subfolder))
        return files
        
    def export_to_csv(self, result: Dict[str, Any], filename: str):
        """Export results to CSV"""
        with open(filename, 'w', newline='', encoding='utf-8') as csvfile:
            writer = csv.writer(csvfile)
            writer.writerow(['Path', 'Name', 'Type', 'Size (bytes)', 'Size (formatted)', 'File Count', 'Folder Count'])
            
            def write_folder_data(folder_data: Dict[str, Any]):
                writer.writerow([
                    folder_data['path'],
                    folder_data['name'],
                    'Folder',
                    folder_data['total_size'],
                    self.client.format_size(folder_data['total_size']),
                    folder_data['file_count'],
                    folder_data['folder_count']
                ])
                
                # Write files
                for file in folder_data['files']:
                    writer.writerow([
                        file['path'],
                        file['name'],
                        'File',
                        file['size'],
                        self.client.format_size(file['size']),
                        0,
                        0
                    ])
                
                # Write subfolders
                for subfolder in folder_data['subfolders']:
                    write_folder_data(subfolder)
            
            write_folder_data(result)
        
        print(f"\n✓ Results exported to: {filename}")
        
    def export_to_json(self, result: Dict[str, Any], filename: str):
        """Export results to JSON"""
        with open(filename, 'w', encoding='utf-8') as jsonfile:
            json.dump(result, jsonfile, indent=2, default=str)
        
        print(f"✓ Results exported to: {filename}")


def main():
    """Main function"""
    parser = argparse.ArgumentParser(
        description='Calculate SharePoint folder sizes using Microsoft Graph API'
    )
    parser.add_argument('--tenant-id', required=True, help='Azure AD tenant ID')
    parser.add_argument('--client-id', required=True, help='Azure AD application client ID')
    parser.add_argument('--client-secret', help='Azure AD application client secret')
    parser.add_argument('--site-url', required=True, help='SharePoint site URL')
    parser.add_argument('--folder-path', required=True, help='Folder path to analyze')
    parser.add_argument('--output-csv', default='folder_sizes.csv', help='Output CSV filename')
    parser.add_argument('--output-json', default='folder_sizes.json', help='Output JSON filename')
    
    args = parser.parse_args()
    
    print("\n🚀 SharePoint Folder Size Calculator")
    print("=" * 60)
    
    # Check for client secret in environment if not provided
    client_secret = args.client_secret or os.environ.get('CLIENT_SECRET')
    
    if not client_secret:
        print("✗ Client secret not provided")
        sys.exit(1)
    
    # Create client
    client = GraphClient(
        tenant_id=args.tenant_id,
        client_id=args.client_id,
        client_secret=client_secret
    )
    
    # Authenticate
    if not client.authenticate():
        print("✗ Authentication failed. Exiting.")
        sys.exit(1)
    
    # Create calculator
    calculator = FolderSizeCalculator(client)
    
    # Analyze folder
    result = calculator.analyze_site(args.site_url, args.folder_path)
    
    if result:
        calculator.print_summary(result)
        calculator.export_to_csv(result, args.output_csv)
        calculator.export_to_json(result, args.output_json)
        print("\n✅ Analysis complete!")
    else:
        print("\n❌ Failed to analyze folder.")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: nodejs/sharepoint-folder-size.js
#!/usr/bin/env node
/**
 * SharePoint Folder Size Calculator with Multiple Authentication Methods
 * This script calculates folder sizes recursively in SharePoint Online
 * using either client credentials or device code flow for authentication.
 */

import { spfi } from "@pnp/sp";
import "@pnp/sp/webs";
import "@pnp/sp/folders";
import "@pnp/sp/files";
import { PublicClientApplication, ConfidentialClientApplication } from "@azure/msal-node";
import { promises as fs } from 'fs';
import { createObjectCsvWriter } from 'csv-writer';
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';
import chalk from 'chalk';
import ora from 'ora';

class AuthProvider {
    constructor(tenantId, clientId, siteUrl, clientSecret = null) {
        this.tenantId = tenantId;
        this.clientId = clientId;
        this.siteUrl = siteUrl;
        this.clientSecret = clientSecret;
        this.authority = `https://login.microsoftonline.com/${tenantId}`;
    }

    async authenticate() {
        if (this.clientSecret) {
            return await this.authenticateClientCredentials();
        } else {
            return await this.acquireTokenDeviceCode();
        }
    }

    async authenticateClientCredentials() {
        console.log('\n' + '='.repeat(50));
        console.log(chalk.bold.yellow('CLIENT CREDENTIALS AUTHENTICATION'));
        console.log('='.repeat(50));
        console.log(chalk.cyan('Authenticating with client secret...'));

        const config = {
            auth: {
                clientId: this.clientId,
                authority: this.authority,
                clientSecret: this.clientSecret
            }
        };

        const cca = new ConfidentialClientApplication(config);
        const tokenRequest = {
            scopes: [`${this.siteUrl}/.default`]
        };

        try {
            const response = await cca.acquireTokenByClientCredential(tokenRequest);
            console.log(chalk.green('✓ Authentication successful!'));
            console.log('='.repeat(50) + '\n');
            return response.accessToken;
        } catch (error) {
            console.error(chalk.red('✗ Authentication failed!'));
            console.error(chalk.red(`Error: ${error.message}`));
            return null;
        }
    }

    async acquireTokenDeviceCode() {
        const config = {
            auth: {
                clientId: this.clientId,
                authority: this.authority
            }
        };
        
        const pca = new PublicClientApplication(config);
        
        const deviceCodeRequest = {
            scopes: [`${this.siteUrl}/.default`],
            deviceCodeCallback: (response) => {
                console.log('\n' + '='.repeat(50));
                console.log(chalk.bold.yellow('DEVICE CODE AUTHENTICATION'));
                console.log('='.repeat(50));
                console.log(chalk.cyan('\nTo sign in, use a web browser to open the page:'));
                console.log(chalk.bold.blue(response.verificationUri));
                console.log(chalk.cyan('\nEnter the code:'), chalk.bold.green(response.userCode));
                console.log(chalk.yellow('\nWaiting for authentication...'));
                console.log('='.repeat(50) + '\n');
            }
        };

        try {
            const response = await pca.acquireTokenByDeviceCode(deviceCodeRequest);
            console.log(chalk.green('✓ Authentication successful!'));
            return response.accessToken;
        } catch (error) {
            console.error(chalk.red('✗ Authentication failed!'));
            console.error(chalk.red(`Error: ${error.message}`));
            return null;
        }
    }
}

class SharePointFolderSizeCalculator {
    constructor(siteUrl, accessToken) {
        this.siteUrl = siteUrl;
        this.sp = spfi(siteUrl).using((instance) => {
            instance.on.auth.replace(async (url, init) => {
                init.headers = {
                    ...init.headers,
                    "Authorization": `Bearer ${accessToken}`,
                };
                return [url, init];
            });
            return instance;
        });
    }

    async getFolderSizeRecursive(folderPath) {
        const spinner = ora(`Analyzing folder: ${folderPath}`).start();
        
        try {
            // Ensure folder path is properly formatted
            if (!folderPath.startsWith('/')) {
                folderPath = '/' + folderPath;
            }

            const result = {
                path: folderPath,
                name: folderPath.split('/').pop() || 'Root',
                totalSize: 0,
                fileCount: 0,
                folderCount: 0,
                files: [],
                subfolders: []
            };

            await this.processFolderRecursive(folderPath, result, spinner);
            
            spinner.succeed(`Analysis complete for: ${folderPath}`);
            return result;

        } catch (error) {
            spinner.fail(`Error accessing folder ${folderPath}`);
            console.error(chalk.red(`Error details: ${error.message}`));
            return null;
        }
    }

    async processFolderRecursive(folderPath, result, spinner, depth = 0) {
        const indent = '  '.repeat(depth);
        
        try {
            // Get folder object
            const folder = this.sp.web.getFolderByServerRelativePath(folderPath);
            
            // Get files in current folder
            const files = await folder.files.select("Name", "Length", "TimeLastModified", "ServerRelativeUrl")();
            
            // Process each file
            for (const file of files) {
                const fileInfo = {
                    name: file.Name,
                    size: file.Length,
                    lastModified: file.TimeLastModified,
                    path: file.ServerRelativeUrl
                };
                
                result.files.push(fileInfo);
                result.totalSize += file.Length;
                result.fileCount++;
                
                spinner.text = `${indent}  📄 ${file.Name} (${this.formatSize(file.Length)})`;
            }

            // Get subfolders
            const subfolders = await folder.folders.filter("Name ne 'Forms'")();
            
            // Process each subfolder
            for (const subfolder of subfolders) {
                // Skip hidden system folders
                if (subfolder.Name.startsWith('_')) {
                    continue;
                }
                
                spinner.text = `${indent}📁 ${subfolder.Name}/`;
                result.folderCount++;
                
                // Create subfolder result
                const subfolderResult = {
                    path: subfolder.ServerRelativeUrl,
                    name: subfolder.Name,
                    totalSize: 0,
                    fileCount: 0,
                    folderCount: 0,
                    files: [],
                    subfolders: []
                };
                
                // Recursively process subfolder
                await this.processFolderRecursive(
                    subfolder.ServerRelativeUrl, 
                    subfolderResult, 
                    spinner, 
                    depth + 1
                );
                
                // Add subfolder results to parent
                result.subfolders.push(subfolderResult);
                result.totalSize += subfolderResult.totalSize;
                result.fileCount += subfolderResult.fileCount;
                result.folderCount += subfolderResult.folderCount;
            }
        } catch (error) {
            spinner.warn(`${indent}⚠️  Error processing folder: ${error.message}`);
        }
    }

    formatSize(bytes) {
        const units = ['B', 'KB', 'MB', 'GB', 'TB'];
        let size = bytes;
        let unitIndex = 0;
        
        while (size >= 1024 && unitIndex < units.length - 1) {
            size /= 1024;
            unitIndex++;
        }
        
        return `${size.toFixed(2)} ${units[unitIndex]}`;
    }

    printSummary(result) {
        if (!result) return;
        
        console.log('\n' + '='.repeat(60));
        console.log(chalk.bold.cyan('FOLDER SIZE SUMMARY'));
        console.log('='.repeat(60));
        console.log(chalk.white(`Path: ${chalk.yellow(result.path)}`));
        console.log(chalk.white(`Total Size: ${chalk.green(this.formatSize(result.totalSize))} (${result.totalSize.toLocaleString()} bytes)`));
        console.log(chalk.white(`Files: ${chalk.blue(result.fileCount.toLocaleString())}`));
        console.log(chalk.white(`Folders: ${chalk.blue(result.folderCount.toLocaleString())}`));
        console.log('='.repeat(60));
        
        // Print top 10 largest files
        const allFiles = this.getAllFiles(result);
        allFiles.sort((a, b) => b.size - a.size);
        
        if (allFiles.length > 0) {
            console.log('\n' + chalk.bold.cyan('Top 10 Largest Files:'));
            console.log('-'.repeat(60));
            
            const top10 = allFiles.slice(0, 10);
            top10.forEach((file, index) => {
                const num = chalk.gray(`${index + 1}.`.padStart(3));
                const name = chalk.white(file.name.padEnd(40));
                const size = chalk.green(this.formatSize(file.size).padStart(10));
                console.log(`${num} ${name} ${size}`);
            });
        }
    }

    getAllFiles(result) {
        let files = [...result.files];
        for (const subfolder of result.subfolders) {
            files = files.concat(this.getAllFiles(subfolder));
        }
        return files;
    }

    async exportToCSV(result, filename) {
        const records = [];
        
        const collectRecords = (folderData) => {
            // Add folder record
            records.push({
                path: folderData.path,
                name: folderData.name,
                type: 'Folder',
                sizeBytes: folderData.totalSize,
                sizeFormatted: this.formatSize(folderData.totalSize),
                fileCount: folderData.fileCount,
                folderCount: folderData.folderCount
            });
            
            // Add file records
            for (const file of folderData.files) {
                records.push({
                    path: file.path,
                    name: file.name,
                    type: 'File',
                    sizeBytes: file.size,
                    sizeFormatted: this.formatSize(file.size),
                    fileCount: 0,
                    folderCount: 0
                });
            }
            
            // Process subfolders
            for (const subfolder of folderData.subfolders) {
                collectRecords(subfolder);
            }
        };
        
        collectRecords(result);
        
        const csvWriter = createObjectCsvWriter({
            path: filename,
            header: [
                { id: 'path', title: 'Path' },
                { id: 'name', title: 'Name' },
                { id: 'type', title: 'Type' },
                { id: 'sizeBytes', title: 'Size (bytes)' },
                { id: 'sizeFormatted', title: 'Size (formatted)' },
                { id: 'fileCount', title: 'File Count' },
                { id: 'folderCount', title: 'Folder Count' }
            ]
        });
        
        await csvWriter.writeRecords(records);
        console.log(chalk.green(`\n✓ Results exported to: ${filename}`));
    }

    async exportToJSON(result, filename) {
        await fs.writeFile(filename, JSON.stringify(result, null, 2));
        console.log(chalk.green(`✓ Results exported to: ${filename}`));
    }
}

async function main() {
    const argv = yargs(hideBin(process.argv))
        .option('tenant-id', {
            describe: 'Azure AD tenant ID',
            type: 'string',
            demandOption: true
        })
        .option('client-id', {
            describe: 'Azure AD application client ID',
            type: 'string',
            demandOption: true
        })
        .option('client-secret', {
            describe: 'Azure AD application client secret (for app-only auth)',
            type: 'string'
        })
        .option('site-url', {
            describe: 'SharePoint site URL',
            type: 'string',
            demandOption: true
        })
        .option('folder-path', {
            describe: 'Folder path to analyze',
            type: 'string',
            demandOption: true
        })
        .option('output-csv', {
            describe: 'Output CSV filename',
            type: 'string',
            default: 'folder_sizes.csv'
        })
        .option('output-json', {
            describe: 'Output JSON filename',
            type: 'string',
            default: 'folder_sizes.json'
        })
        .help()
        .argv;

    console.log(chalk.bold.magenta('\n🚀 SharePoint Folder Size Calculator'));
    console.log('='.repeat(60));

    // Check for client secret in environment if not provided
    const clientSecret = argv.clientSecret || process.env.CLIENT_SECRET;

    // Authenticate
    const authProvider = new AuthProvider(
        argv.tenantId,
        argv.clientId,
        argv.siteUrl,
        clientSecret
    );

    const accessToken = await authProvider.authenticate();
    if (!accessToken) {
        console.error(chalk.red('✗ Authentication failed. Exiting.'));
        process.exit(1);
    }

    // Calculate folder sizes
    const calculator = new SharePointFolderSizeCalculator(argv.siteUrl, accessToken);
    
    const startTime = Date.now();
    const result = await calculator.getFolderSizeRecursive(argv.folderPath);
    const endTime = Date.now();

    if (result) {
        calculator.printSummary(result);
        await calculator.exportToCSV(result, argv.outputCsv);
        await calculator.exportToJSON(result, argv.outputJson);
        
        const duration = ((endTime - startTime) / 1000).toFixed(2);
        console.log(chalk.cyan(`\n⏱️  Processing time: ${duration} seconds`));
        console.log(chalk.green.bold('\n✅ Analysis complete!'));
    } else {
        console.error(chalk.red.bold('\n❌ Failed to analyze folder.'));
        process.exit(1);
    }
}

// Run the main function
main().catch(error => {
    console.error(chalk.red('Unexpected error:'), error);
    process.exit(1);
});